xmltools/xml2utf8.1
.\" Copyright (c) 2002 Bill C. Riemers
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2 of
.\" the License, or (at your option) any later version.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual. Otherwise check the web site
.\" of the Free Software Foundation at http://www.fsf.org.
.\"
.\" I, Bill C. Riemers, hereby grant all rights to this code,
.\" provided usage complies with the GPL or a written exception to 
.\" the GPL granted by any of Bill C. Riemers, Leon Bottou, 
.\" Yann Le Cun, or the Free Source Foundation.
.\"
.\" As derived work,  the following LizardTech, Inc. copyright 
.\" notice applies to this code.
.\"  ------------------------------------------------------------------
.\" DjVuLibre-3.5 is derived from the DjVu(r) Reference Library
.\" distributed by Lizardtech Software.  On July 19th 2002, Lizardtech 
.\" Software authorized us to replace the original DjVu(r) Reference 
.\" Library notice by the following text (see doc/lizard2002.djvu):
.\"
.\"  ------------------------------------------------------------------
.\" | DjVu (r) Reference Library (v. 3.5)
.\" | Copyright (c) 1999-2001 LizardTech, Inc. All Rights Reserved.
.\" | The DjVu Reference Library is protected by U.S. Pat. No.
.\" | 6,058,214 and patents pending.
.\" |
.\" | This software is subject to, and may be distributed under, the
.\" | GNU General Public License, Version 2. The license should have
.\" | accompanied the software or you may obtain a copy of the license
.\" | from the Free Software Foundation at http://www.fsf.org .
.\" |
.\" | The computer code originally released by LizardTech under this
.\" | license and unmodified by other parties is deemed "the LIZARDTECH
.\" | ORIGINAL CODE."  Subject to any third party intellectual property
.\" | claims, LizardTech grants recipient a worldwide, royalty-free, 
.\" | non-exclusive license to make, use, sell, or otherwise dispose of 
.\" | the LIZARDTECH ORIGINAL CODE or of programs derived from the 
.\" | LIZARDTECH ORIGINAL CODE in compliance with the terms of the GNU 
.\" | General Public License.   This grant only confers the right to 
.\" | infringe patent claims underlying the LIZARDTECH ORIGINAL CODE to 
.\" | the extent such infringement is reasonably necessary to enable 
.\" | recipient to make, have made, practice, sell, or otherwise dispose 
.\" | of the LIZARDTECH ORIGINAL CODE (or portions thereof) and not to 
.\" | any greater extent that may be necessary to utilize further 
.\" | modifications or combinations.
.\" |
.\" | The LIZARDTECH ORIGINAL CODE is provided "AS IS" WITHOUT WARRANTY
.\" | OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
.\" | TO ANY WARRANTY OF NON-INFRINGEMENT, OR ANY IMPLIED WARRANTY OF
.\" | MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
.\" +------------------------------------------------------------------
.TH XML2UTF8 1 "11/15/2002" "DjVuLibre-3.5" "DjVuLibre-3.5"
.de SS
.SH \\0\\0\\0\\$*
..
.SH NAME
xml2utf8 \- Converts a Extensible Markup Language (XML) document to UTF8 character encoding.

.SH SYNOPSIS
.BI "xml2utf8 " "inputxmlfile" " " "outputxmlfile" " "

.SH DESCRIPTION
Program 
.B xml2utf8
converts an Extensible Markup Language (XML) file to 
.SM UTF8 
character encoding.

.SH REMARKS
This conversion is needed to display 
.SM XML 
files in old versions of Windows Notepad and many other Windows programs.

.SH CREDITS
This program was written by Bill C. Riemers <docbill@sourceforge.net>.

.SH SEE ALSO
.BR djvutoxml (1),
.BR djvuxmlparser (1),
.BR djvuxml-s (3),
.BR xml2native (1),
.BR xml2utf16 (1),
and
.BR utf8 (3).



NEWS

NEW IN VERSION 3.5.30
------------------------
- various bug fixes


NEW IN VERSION 3.5.29
------------------------
- various bug fixes
- additional tests for corrupted files
- fixes for clang warnings


NEW IN VERSION 3.5.28
------------------------
- ddjvu: tiff generation improvements
- djvumake: security checks on INCL chunks
- all: updated for modern compilers
- win: moved to msvc2019
- osx: nfd filename normalization
- bugs: fixed several crashes on invalid inputs
- miniexp: fixed escape printout and macrochars
- djvudigital: can use poppler to find text
- csepdjvu: handle T comments for page titles
- bytestream: fixed 2GB limit
- gexception, gthread: cleanup obsolete code


NEW IN VERSION 3.5.27
------------------------
- simplified configuration scripts
- deadlock fixes
- miniexp under win32 uses tlsalloc instead of _thread vars (pb under xp)
- djvudigital can extract text and metadata using pdftotext.
- duplicate page titles are accepted
- bug fixes: overflow in bitmap/pixmap/iw44image sizes
- big fixes: handling INCL chunks in djvumake
- bug fixes: better support for large files (2gb) 


NEW IN VERSION 3.5.26
------------------------
- Thread safe miniexp
- Now using the standard autotools approach (thanks to Vincent Torri).
- New windows installer (thanks to Konstantin Kravtsov).
- Fixed Russian code page issues in windows.
- General speedup thanks to lock-free smart pointers.
- ddjvu can produce one file per page.
- djvused can now set the default page orientation.
- Bug fixes all around.


NEW IN VERSION 3.5.25
---------------------
- Removed the deprecated djview3 code.
- Removed the deprecated cothreads code.
- Removed the unmaintained japanese man pages.
- Added ddjvu option to skip corrupted pages.
- Miniexp API is now reentrant.
- Bug fixes all around.


NEW IN VERSION 3.5.24
---------------------
- Set ddjvuapi_version to 20.
- Added ddjvuapi support for changing the white point.
- Added option "-u" in djvused.
- Added option "-o djvufile" in djvuxmlparser.
- Added support for <CHARACTER> tag in djvuxmlparser.
- Added print-xmp, set-xmp, remove-smp in djvused.


NEW IN VERSION 3.5.23
---------------------
- Djview3 has been deprecated by djview4 whose source comes in 
  its own tarball. Use configure option --enable-djview to force
  the compilation of djview3.
- Set ddjvuapi_version to 19.
- Added ddjvu_document_create_by_filename_utf8
- Generate icons with rsvg rather than imagemagick.
- djvused: correctly outputs escape characters.
- djvumake: added options to generate foreground colors.
- Various bug fixes.

NEW IN VERSION 3.5.22
---------------------
- Probably the last release before djvulibre-4
- Backported djview4 version of nsdejavu.so 
- Various bug fixes

NEW IN VERSION 3.5.21
---------------------
- Speed improvement on multicore platforms
- Ddjvu now can produce pdf output.
- Various bug fixes

NEW IN VERSION 3.5.20
---------------------
- Csepdjvu handle gsdjvu hyperlink comments.
- Djvudigital therefore handles hyperlinks.

NEW IN VERSION 3.5.19
---------------------
- License change. Starting with version 3.5.19, 
  DjVuLibre is distributed under either version 2 of the GPL
  or (at your option) any later version.
  Previous versions were released strictly under 
  the GPL version 2.

NEW IN VERSION 3.5.18
---------------------
- Renamed djview as djview3.
- Various enhancement of DDJVUAPI.
- Partial support for page titles instead of page numbers.
- Clarification of metadata information.
- Fixed djvumake mask separation feature.
- Complain when given a so-called "secure" djvu file.
- Fixed handling of urls with cgi arguments
- Fixed bug related to caching and ddjvuapi
- Reorganized ${datadir}/djvu
- Provide scripts in ${datadir}/djvu/*/desktop to register
  the djvu mime types and the djview menu entries.

NEW IN VERSION 3.5.17
---------------------
- Various enhancement of DDJVUAPI
- Compiles with GCC-4.1.
- Compiles with the SunCC compiler.
- Command line utilities compile under WIN32 
  using MSVC using the provided project files.
- Minor bug fixes in csepdjvu text extraction.
- Fixed leak in csepdjvu, cpaldjvu and cjb2.

NEW IN VERSION 3.5.16
---------------------
- Shared library build system updates.

NEW IN VERSION 3.5.15
---------------------
- Fixes for compilation with GCC-4.0.

NEW IN VERSION 3.5.14
---------------------
- Fixes for compilation with GCC-3.4.
- Cjb2 lossy mode vastly improved (thanks to Ilya Mezhirov)
- Revamped install for desktop information files
- Compilation uses libtool and makes shared libs on more platforms.
- Uniform decoding api ddjvuapi.h.
- Enhanced ddjvu based on ddjvuapi.
- ddjvu can produce tiff files.

NEW IN VERSION 3.5.13
---------------------
- Navigation history in djview.
- New option -fullscreen in djview.
- Silently processes files up to version 25.
- Only produces files with version 24.
- French and Japanese localization.
- Japanese man pages.
- Fixes for MacOS.
- Compilation of XMLTools enabled by default.
- KDE and GNOME integration
- CJB2 now handles tiff files.

NEW IN VERSION 3.5.12
---------------------
- Improved color quantization (fix cpaldjvu artifacts)
- Fixed handling of non latin characters.
- Worked around problems with gcc-3.3.1.

NEW IN VERSION 3.5.11
---------------------
- All private symbols in DJVU namespace
- Fixed keyboard focus in kde-3.1.1
- Fixed toolbar control arguments.
- Incremented djvuversion to 24 (LT again)
- Fixed djvm.

NEW IN VERSION 3.5.10
---------------------
- Added support for mouse wheel.
- Added support for LT toolbar control args.
- Added DESTDIR in Makefiles.
- Incremented djvuversion to 23.
- Djview key space now does continuous reading
- Fixed thumbnail generation on empty pages.
- Improved csepdjvu option -bgwhite.
- Djview and djvups now can print booklets.
- Djview main dialogs have a help button.
- Annotation strings now recognize C escape sequences.

NEW IN VERSION 3.5.9
--------------------
- Fix for gcc-3.2.
- Fix for broken command ddjvu.

xmltools/xml2utf8.cpp
//C-  -*- C++ -*-
//C- -------------------------------------------------------------------
//C- Xmltools for DjVu3 and DjVuLibre-3.5
//C- Copyright (c) 2002  Bill C. Riemers
//C- -------------------------------------------------------------------
//C- This software is subject to, and may be distributed under, the
//C- GNU General Public License, Version 2. The license should have
//C- accompanied the software or you may obtain a copy of the license
//C- from the Free Software Foundation at http://www.fsf.org .
//C-
//C- This program is distributed in the hope that it will be useful,
//C- but WITHOUT ANY WARRANTY; without even the implied warranty of
//C- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//C- GNU General Public License for more details.
//C- 
//C- I, Bill C. Riemers, hereby grant all rights to this code,
//C- provided usage complies with the GPL or a written exception to 
//C- the GPL granted by any of Bill C. Riemers, Leon Bottou, 
//C- Yann Le Cun, or the Free Source Foundation.
//C-
//C- ------------------------------------------------------------------
//C- DjVuLibre-3.5 is derived from the DjVu(r) Reference Library
//C- distributed by Lizardtech Software.  On July 19th 2002, Lizardtech 
//C- Software authorized us to replace the original DjVu(r) Reference 
//C- Library notice by the following text (see doc/lizard2002.djvu):
//C-
//C-  ------------------------------------------------------------------
//C- | DjVu (r) Reference Library (v. 3.5)
//C- | Copyright (c) 1999-2001 LizardTech, Inc. All Rights Reserved.
//C- | The DjVu Reference Library is protected by U.S. Pat. No.
//C- | 6,058,214 and patents pending.
//C- |
//C- | This software is subject to, and may be distributed under, the
//C- | GNU General Public License, Version 2. The license should have
//C- | accompanied the software or you may obtain a copy of the license
//C- | from the Free Software Foundation at http://www.fsf.org .
//C- |
//C- | The computer code originally released by LizardTech under this
//C- | license and unmodified by other parties is deemed "the LIZARDTECH
//C- | ORIGINAL CODE."  Subject to any third party intellectual property
//C- | claims, LizardTech grants recipient a worldwide, royalty-free, 
//C- | non-exclusive license to make, use, sell, or otherwise dispose of 
//C- | the LIZARDTECH ORIGINAL CODE or of programs derived from the 
//C- | LIZARDTECH ORIGINAL CODE in compliance with the terms of the GNU 
//C- | General Public License.   This grant only confers the right to 
//C- | infringe patent claims underlying the LIZARDTECH ORIGINAL CODE to 
//C- | the extent such infringement is reasonably necessary to enable 
//C- | recipient to make, have made, practice, sell, or otherwise dispose 
//C- | of the LIZARDTECH ORIGINAL CODE (or portions thereof) and not to 
//C- | any greater extent that may be necessary to utilize further 
//C- | modifications or combinations.
//C- |
//C- | The LIZARDTECH ORIGINAL CODE is provided "AS IS" WITHOUT WARRANTY
//C- | OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
//C- | TO ANY WARRANTY OF NON-INFRINGEMENT, OR ANY IMPLIED WARRANTY OF
//C- | MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE.
//C- +------------------------------------------------------------------
// 
// $Id$
// $Name$

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif
#if NEED_GNUG_PRAGMAS
# pragma implementation
#endif

/** @name xml2utf8

    {\bf Synopsis}
    \begin{verbatim}
        xml2utf8 [<inputfile>] [<outputfile>]
    \end{verbatim}

    @author
    Dr Bill C Riemers <docbill@sourceforge.net>
    @version
    #$Id$# */
//@{
//@}




#include "GException.h"
#include "UnicodeByteStream.h"
#include "ByteStream.h"
#include "DjVuMessage.h"
#include "GURL.h"

#include <locale.h>
#include <stdlib.h>
#include <stdio.h>

void
usage(const GUTF8String &name)
{
  DjVuPrintErrorUTF8(
          "%s -- xml to UTF8 MBS encoding conversion\n"
          "  Copyright © 2001 LizardTech, Inc. All Rights Reserved.\n"
          "Usage: %s <infile> <outfile>\n",(const char *)name,(const char *)name);
}

int 
main(int argc, char **argv)
{
  setlocale(LC_ALL,"");
  djvu_programname(argv[0]);
  GArray<GUTF8String> dargv(0,argc-1);
  for(int i=0;i<argc;++i)
    dargv[i]=GNativeString(argv[i]);
  if(argc < 3 || dargv[1] == "-help" || dargv[1] == "--help")
  {
    usage(GURL::Filename::UTF8(dargv[0]).fname());
    exit(0);
  }
  GURL::Filename::UTF8 inurl((argc<2)?GUTF8String("-"):dargv[1]);
  GURL::Filename::UTF8 outurl((argc<3)?GUTF8String("-"):dargv[argc-1]);
  G_TRY
  {
    GP<ByteStream> bs(ByteStream::create(inurl,"rb")); 
    {
      GP<XMLByteStream> uni=XMLByteStream::create(bs);
      bs=ByteStream::create();
      GUTF8String ustr;
      while((ustr=uni->gets()).length())
      {
        bs->writestring(ustr);
      }
    }
    bs->seek(0L);
    GP<ByteStream> outbs=ByteStream::create(outurl,"wb");
    outbs->write24(0xEFBBBF);
    outbs->copy(*bs);
  }
  G_CATCH(ex)
  {
    ex.perror();
    exit(1);
  }
  G_ENDCATCH;
  exit(0);
#ifdef WIN32
  return 0;
#endif
}


